[0.10.0] panic: interface conversion: interface is float64, not int64

Version:

```
vagrant@influx1:~$ influxd version
InfluxDB v0.10.0.n1453795235 (git: master 9d3c9329a69c1dbd2ed7058acd8c2711c781d571, built 2016-01-26T08:01:53.793059)
```

Query and trace:

```
[query] 2016/01/29 17:14:04 SELECT last(seriesCreate) FROM _internal.monitor."shard" WHERE path =~ /.*test.*/
panic: interface conversion: interface is float64, not int64

goroutine 274 [running]:
github.com/influxdb/influxdb/tsdb.greaterThan(0x9b1bc0, 0xc208496d60, 0x9b0c40, 0xc208496bc0, 0xc20822f700)
        /tmp/tmp.QSk2v30WJm/src/github.com/influxdb/influxdb/tsdb/functions.go:1723 +0x1d1
github.com/influxdb/influxdb/tsdb.ReduceLast(0xc20822f6a0, 0x2, 0x2, 0x0, 0x0)
        /tmp/tmp.QSk2v30WJm/src/github.com/influxdb/influxdb/tsdb/functions.go:1086 +0x39b
github.com/influxdb/influxdb/tsdb.(*AggregateExecutor).execute(0xc20822e9e0, 0xc2099d6f00, 0xc2099d6c60)
        /tmp/tmp.QSk2v30WJm/src/github.com/influxdb/influxdb/tsdb/aggregate.go:144 +0x1449
created by github.com/influxdb/influxdb/tsdb.(*AggregateExecutor).Execute
        /tmp/tmp.QSk2v30WJm/src/github.com/influxdb/influxdb/tsdb/aggregate.go:48 +0x64
```

Interestingly, I ran the repro steps very fast and the server didn't crash:

```
[query] 2016/01/29 18:02:46 SELECT last(seriesCreate) FROM _internal.monitor."shard" WHERE path =~ /.*test.*/
[http] 2016/01/29 18:02:46 127.0.0.1 - - [29/Jan/2016:18:02:46 +0000] GET /query?db=_internal&epoch=ns&q=select+last%28seriesCreate%29+from+%22shard%22+where+path+%3D~+%2F.%2Atest.%2A%2F HTTP/1.1 200 40 - InfluxDBShell/0.10.0.n1453795235 80506ade-c6b2-11e5-8058-000000000000 4.952749ms
[query] 2016/01/29 18:02:50 SELECT last(seriesCreate) FROM _internal.monitor."shard" WHERE path =~ /.*test.*/
[http] 2016/01/29 18:02:50 127.0.0.1 - - [29/Jan/2016:18:02:50 +0000] GET /query?db=_internal&epoch=ns&q=select+last%28seriesCreate%29+from+%22shard%22+where+path+%3D~+%2F.%2Atest.%2A%2F HTTP/1.1 200 40 - InfluxDBShell/0.10.0.n1453795235 82a7cc0c-c6b2-11e5-8059-000000000000 3.892318ms
[cluster] 2016/01/29 18:02:51 accept remote connection from 172.28.128.68:50219
[query] 2016/01/29 18:02:59 SELECT seriesCreate FROM _internal.monitor."shard" WHERE path =~ /.*test.*/
[http] 2016/01/29 18:02:59 127.0.0.1 - - [29/Jan/2016:18:02:59 +0000] GET /query?db=_internal&epoch=ns&q=select+seriesCreate+from+%22shard%22+where+path+%3D~+%2F.%2Atest.%2A%2F HTTP/1.1 200 115 - InfluxDBShell/0.10.0.n1453795235 87f64c37-c6b2-11e5-805a-000000000000 3.993275ms
[query] 2016/01/29 18:03:05 SELECT seriesCreate FROM _internal.monitor."shard" WHERE path =~ /.*test.*/
[http] 2016/01/29 18:03:05 127.0.0.1 - - [29/Jan/2016:18:03:05 +0000] GET /query?db=_internal&epoch=ns&q=select+seriesCreate+from+%22shard%22+where+path+%3D~+%2F.%2Atest.%2A%2F HTTP/1.1 200 118 - InfluxDBShell/0.10.0.n1453795235 8bc93e2e-c6b2-11e5-805b-000000000000 3.73771ms
[query] 2016/01/29 18:03:07 SELECT seriesCreate FROM _internal.monitor."shard" WHERE path =~ /.*test.*/
[http] 2016/01/29 18:03:07 127.0.0.1 - - [29/Jan/2016:18:03:07 +0000] GET /query?db=_internal&epoch=ns&q=select+seriesCreate+from+%22shard%22+where+path+%3D~+%2F.%2Atest.%2A%2F HTTP/1.1 200 118 - InfluxDBShell/0.10.0.n1453795235 8ce94ac5-c6b2-11e5-805c-000000000000 5.591682ms
[query] 2016/01/29 18:03:16 SELECT last(seriesCreate) FROM _internal.monitor."shard" WHERE path =~ /.*test.*/
panic: interface conversion: interface is float64, not int64
```

Only after this log message:

```
[cluster] 2016/01/29 18:02:51 accept remote connection from 172.28.128.68:50219
```

Does the query crash the service.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -33,6 +33,7 @@
 - [#5469](https://github.com/influxdata/influxdb/issues/5469): Conversion from bz1 to tsm doesn't work as described
 - [#5449](https://github.com/influxdata/influxdb/issues/5449): panic when dropping collectd points
 - [#5455](https://github.com/influxdata/influxdb/issues/5455): panic: runtime error: slice bounds out of range when loading corrupted wal segment
+- [#5478](https://github.com/influxdata/influxdb/issues/5478): panic: interface conversion: interface is float64, not int64
 
 ## v0.9.6 [2015-12-09]
 

--- a/tsdb/functions.go
+++ b/tsdb/functions.go
@@ -1718,28 +1718,28 @@ func (a rawOutputs) Less(i, j int) bool { return a[i].Time < a[j].Time }
 func (a rawOutputs) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
 
 func greaterThan(a, b interface{}) bool {
-	switch t := a.(type) {
+	switch a := a.(type) {
 	case int64:
-		switch bb := b.(type) {
+		switch b := b.(type) {
 		case int64:
-			return t > bb
+			return a > b
 		case float64:
-			return float64(t) > bb
+			return float64(a) > b
 		}
 	case float64:
-		switch bb := b.(type) {
+		switch b := b.(type) {
 		case int64:
-			return t > float64(bb)
+			return a > float64(b)
 		case float64:
-			return t > bb
+			return a > b
 		}
 	case string:
-		switch bb := b.(type) {
+		switch b := b.(type) {
 		case string:
-			return t > bb
+			return a > b
 		}
 	case bool:
-		return t == true
+		return a == true
 	}
 	return false
 }

--- a/tsdb/functions.go
+++ b/tsdb/functions.go
@@ -1720,11 +1720,24 @@ func (a rawOutputs) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
 func greaterThan(a, b interface{}) bool {
 	switch t := a.(type) {
 	case int64:
-		return t > b.(int64)
+		switch bb := b.(type) {
+		case int64:
+			return t > bb
+		case float64:
+			return float64(t) > bb
+		}
 	case float64:
-		return t > b.(float64)
+		switch bb := b.(type) {
+		case int64:
+			return t > float64(bb)
+		case float64:
+			return t > bb
+		}
 	case string:
-		return t > b.(string)
+		switch bb := b.(type) {
+		case string:
+			return t > bb
+		}
 	case bool:
 		return t == true
 	}

--- a/tsdb/functions_test.go
+++ b/tsdb/functions_test.go
@@ -1,6 +1,7 @@
 package tsdb
 
 import (
+	"math"
 	"reflect"
 	"testing"
 	"time"
@@ -991,4 +992,44 @@ func TestInitializeUnmarshallerTopBottom(t *testing.T) {
 			t.Errorf("Wrong output. \nexp\n %v\ngot\n %v", spew.Sdump(test.output), spew.Sdump(output))
 		}
 	}
+}
+
+func TestGreaterThan(t *testing.T) {
+
+	for _, tt := range []struct {
+		a, b     interface{}
+		expected bool
+	}{
+		// a = int, b = float
+		{int64(1), float64(2), false},
+		{int64(2), float64(1), true},
+		// a = float, b = int
+		{float64(1), int64(2), false},
+		{float64(2), int64(1), true},
+		// float > float
+		{float64(1), float64(2), false},
+		{float64(2), float64(1), true},
+		// int > int
+		{int64(1), int64(2), false},
+		{int64(2), int64(1), true},
+		// precision at high end of int64 range
+		{int64(math.MaxInt64), int64(math.MaxInt64 - 1), true},
+		{int64(math.MaxInt64 - 1), int64(math.MaxInt64), false},
+		// precision at low end of int64 range
+		{int64(math.MinInt64 + 1), int64(math.MinInt64), true},
+		{int64(math.MinInt64), int64(math.MinInt64 + 1), false},
+		// loss of precision
+		{float64(math.MaxInt64 + 255), int64(math.MaxInt64), false},
+		// smallest float64 delta
+		{math.SmallestNonzeroFloat64, int64(0), true},
+		// comparison at the far reaches of the float64 range
+		{float64(math.MaxFloat64), math.Nextafter(math.MaxFloat64, math.Inf(-1)), true},
+		{math.Nextafter(-math.MaxFloat64, math.Inf(1)), float64(-math.MaxFloat64), true},
+	} {
+		got := greaterThan(tt.a, tt.b)
+		if got != tt.expected {
+			t.Errorf("greaterThan failed for %#T(%[1]v) > %#T(%[2]v), got %v, expected %v.", tt.a, tt.b, got, tt.expected)
+		}
+	}
+
 }

